Boardsizeranked + komiranked  + timecontrolranked + speedranked + human readable (boardsize+komi) in reject messages + UHMAEAT 2.2 !!

wonderingabout opened the pull request:
continuation of #124, this PR is on top of the first one

splitting the old PR for easier review

general look here : https://github.com/wonderingabout/gtp2ogs/tree/boardsizeranked-komiranked-timecontrolranked-speedranked

easy review here : https://github.com/online-go/gtp2ogs/pull/129/files?utf8=%E2%9C%93&diff=split

changes : 

- added many ranked-unranked new options : for speed, timecontrol, boardsize, komi : all now have ranked and unranked

these can be useful in many cases, for example :
- if bot admins has a solid 19x19 bot which can also play 9x9
and/or 13x13 but unreliable, then bot admin may want to allow the
bot to play 19x19 in ranked only, and 9x9 + 13x13 + 19x19 in unranked
- similar case if bot admin wants ranked games to have strict
time mangement like only byoyomi and fischer, but is open to
allowing absolute and/or canadian and/or none in unranked
- similar case for komi, because komi depends on boardsize
- for speed now, maybe bot admin wants ranked games to be all live,
but is open to having unranked games being blitz and/or correspondence too

then the other changes : 

- square boardsize is now human readable thanks to @Dorus 's code in https://github.com/online-go/gtp2ogs/pull/129/commits/2ab884342c38ddac75bc51faf3f770567fa5c239 : `9,13,19` -> `9x9, 13x13, 19x19` , `9,10,11,18,24` -> `9x9, 10x10, 11x11, 18x18, 24x24`
- minor readme enhancements using `npm install` to check for missing packages, other minor fixes
- fix 3 of same family for the `allowed_` family : added a "family guard" for ranked and unranked games or always rejects games
- `null`/`auto` komi displayed as `Automatic` in reject message (requires to now use `--komi Automatic` instead of `--komi auto`

final big change :  
upgraded UHMAEAT to v2.2 !!! fixed and enhanced UHMAEAT for maintimes and periodtimes !!!

big credit goes to gedit "find and replace" feature for helping me do this so easily and reliably !!!
(it can change a block of many lines at the same time in all different parts of the file !!! 
- fixed `undefined` in conn_log
- display requested time by user when rejected (example : "user wanted 1 seconds time, min is 30 seconds"
- fixed canadian period time conn_log (period time for 10 stones is x seconds for all the 10 stones)
- now canadian period time uses the same variables as fischer and co,
so a lot less code maintainance (removed this.canadianSentence, not
needed anymore)
- fix the "duplicate" in reject message "for all the 25 stones for
all the 25 stones
- renamed "main" to "period" for the UHMAEAT for periodtimes
- minor tweaks
- while there is no change, also bumped UHMAEAT for maintimes to
UHMAEAT 2.2 for maintimes too !! more versions more fun !!!

this is a great idea !!


ready to merge now, tested on beta ogs to work !

wonderingabout commented:
@Dorus 

you may want to "textify" this is a more human readable way (extract in `9,13,19,etc` numbers one by one, and print `9x9 13x13 19x19 etcxetc`)

in this code 

```
        if (!config.allowed_sizes[notification.width] && !config.allow_all_sizes && !config.allow_custom_sizes && !config.boardsizeranked && !config.boardsizeunranked) {
            conn_log("square board size " + notification.width + "x" + notification.height + " is not an allowed size, rejecting challenge");
            return { reject: true, msg: "Board size " + notification.width + "x" + notification.height + " is not allowed, please choose one of the allowed square board sizes (same width and height, for example if allowed boardsizes are 9,13,19, it means you can play only 9x9 , 13x13, and 19x19), these are the allowed square board sizes : " + config.boardsize };
        }
```

new reject message would then be 

`
"Board size " + notification.width + "x" + notification.height + " is not allowed, please choose one of the allowed square board sizes : " + n1 + "x" + n1 + "," +  n2 + "x" + n2 + "," + n3 + "x" + n3 + "," + etc + "x" + etc;
`

wonderingabout commented:
reminder for myself

minor thing remaining to finish 

**TODO : fix UHMAEAT v2 canadian periodtime reject message**

wonderingabout commented:
didnt test this PR yet

wonderingabout commented:
solved by @Dorus 

```
  .split(',')
  .map(e => e.trim())
  .map(e => `${e}x${e}`)
  .join(', ')
```

![consoleee](https://user-images.githubusercontent.com/38690718/52749919-f291b980-2fea-11e9-82a2-14942593ac64.png)

Dorus commented on docs/CUSTOM-BRANCHES.md:
````diff
@@ -0,0 +1,80 @@
+You also need to install split2 to use this branch 
+(as admin on windows, as sudo on linux) :
+
+```npm install -g split2```
````
I think just `npm install` install will work since i fixed up packages.json

wonderingabout replied:
indeed but you never know, so i put it like that for now

Dorus replied:
You do know. Also installing it global is rediculous.

wonderingabout replied:
i dont mind doing it if you're 100% sure there wont be any exception, 

and if we are sure about, then shouldnt we also delete this from gtp2ogs readme ? 

https://github.com/online-go/gtp2ogs#3-optional--install-any-missing-nodejs-packages

> 3. Optional : install any missing node.js packages
> This step can be skipped unless you have issues or bugs

> you may need to install extra tools if the Most common usage below fails, such as (run as sudo > for linux, and as admin for windows)

> npm install -g socket.io-client
> npm install -g optimist
> npm install -g tracer

Dorus replied:
Yes, as long as they are present in packages.json,you only need to type `npm install`.

To do so you can use `npm install xxx - -save-prod`

wonderingabout replied:
ok, so `npm install gtp2ogs - save-prod` ? the weird space is at the right place ? (im becoming spaces fanatic too :rofl: )

Dorus replied:
1) No i fixed the space.
2) This is only necessarily to update `packages.json`, once that's done and checked in, others can run just `npm install`.

wonderingabout replied:
ok, all clear

so the gtp2ogs readme will also serve your branch since you share the same readme, all the more reason to fix it !

new readme : 

> in case it fails, run `npm install gtp2ogs - save-prod` to update packages to install list in packages.json
then run : `npm install` to install all missing packages

correct @Dorus ?

if i'm understanding right, then its a nice and very efficient fix to the readme (and to all custom branches at the same time)

Dorus replied:
no no no no no just no.

Only devs should run `--save-prod` to updat `package.json`. Users only need to call `npm install`

wonderingabout replied:
ok got it, gonna commit it !

wonderingabout replied:
fixed in https://github.com/online-go/gtp2ogs/pull/129/commits/02db84a9669deadf2041a3e52e2e62df9b0b593a

Dorus commented on config.js:
```diff
@@ -312,23 +304,27 @@ if (argv.id) {
-if (argv.minrankedhandicap) {
-    console.log("Warning: --minrankedhandicap argument is no longer supported. Use --minhandicapranked instead.");
+if (argv.minrankedhandicap || argv.minunrankedhandicap || argv.maxrankedhandicap || argv.maxunrankedhandicap) {
+    console.log("Warning: --min/max*+/-ranked/unranked*handicap argument is no longer supported.\nUse --min/max*handicap*+/-ranked/unranked instead.");
```
I liked it when it gave one error per modified argument. 

Also, i see a pattern for a function here.

```
function checkParam(b, old, new) {
  if (b) {
    console.log(`Warning: --${old} argument is no longer supported.\nUse --${new} instead.`);
  }
}
```

You can even declare a list here and run the function many times:

```
[[argv.minrankedhandicap, "minrankedhandicap", "minhandicapranked"],
[argv.maxrankedhandicap, "maxrankedhandicap", "maxhandicapranked"]
[argv.minunrankedhandicap, "minunrankedhandicap", "minhandicapunranked"],
[argv.maxunrankedhandicap, "maxunrankedhandicap", "maxhandicapunranked"]]
  .forEach(ar=>checkParam(...ar))
```

wonderingabout replied:
interesting too, but first priority is to fix not working things or add new features :+1: 

but always thanks for teaching me and everyone things !!

Dorus commented on config.js:
```diff
@@ -389,11 +386,11 @@ if (argv.botid || argv.bot || argv.id || argv.minrankedhandicap || argv.maxranke
 
     if (argv.banunranked) {
         for (let i of argv.banunranked.split(',')) {
```
Consider a different variable name here. `i` is not very informative, and usually used for integers, here it's a string, or actually a username.

wonderingabout replied:
i used the same as was here

indeed, `i` is used for counters :+1: , even i know that :rofl: 

out of curiosity, what name should be used in that case ?

Dorus commented on config.js:
```diff
@@ -389,11 +386,11 @@ if (argv.botid || argv.bot || argv.id || argv.minrankedhandicap || argv.maxranke
+            exports.banned_unranked_users[i] = true;
         }
     }
 
```
Would it make sense to do something like 
```
exports.banned_unranked_users = exports.banned_unranked_users.concat(banned_users)
```
?

That way you would no longer need to check both lists later on.

wonderingabout replied:
interesting too, but first priority is to fix not working things or add new features :+1:

roy7 replied:
This has been fixed in devel or maybe from Big to Small PR 1, I don't know.

Dorus commented on config.js:
```diff
@@ -389,11 +386,11 @@ if (argv.botid || argv.bot || argv.id || argv.minrankedhandicap || argv.maxranke
     }
 
-    if (argv.minrank) {
+    if (argv.minrank && !argv.minrankranked && !argv.minrankunranked) {
```
I have some doubts about ignoring `minrank` if `minrankranked` is set. Why not just do both checks if the admin decide to have both set? Less code and arguably more flexible settings.

wonderingabout replied:
because let's imagine bot admin is careless and does : 

- minrank : 15k
- minrankranked : 1d
- minrankunranked 15k

then all games will allow 15k+, while bot admin wanted 1d+ for ranked, and 15k+ for unranked

this adds reliability to gtp2ogs in case bot admin is a bit careless or not knowledgeable about gtp2ogs

Dorus commented on config.js:
```diff
@@ -415,7 +412,51 @@ if (argv.botid || argv.bot || argv.id || argv.minrankedhandicap || argv.maxranke
+        } else {
+            console.error("Could not parse minrankranked " + argv.minrankranked);
+            process.exit();
+        }
```
Consider making a function of the rank parse code.
```
if (argv.minrankranked) {
  try {
  [exports.minrankranked, exports.proonly] = parseRank(argv.minrank);
  } catch (e) {
    if (e === Invalid) {
        console.error("Invalid minrankranked " + argv.minrankranked);
    } else {
        console.error("Could not parse minrankranked " + argv.minrankranked);
    }
    process.exit();
  }
}
```

```
parseRank(rank) {
    let re = /(\d+)([kdp])/;
    let results = argv.minrank.toLowerCase().match(re);
    if (results) {
        if (results[2] == "k") {
            return [30 - parseInt(results[1]), false];
        } else if (results[2] == "d") {
            return [30 - 1 + parseInt(results[1]), false];
        } else if (results[2] == "p") {
            return [36 + parseInt(results[1]), true];
        } else {
            throw "Invalid";
        }
    } else {
        throw "Parse";
    }
}
```

Dorus replied:
I wrote this code really fast, so i hope it works. Please test. If it doesnt work, let me know or just fix it yourself :D

wonderingabout replied:
same as other comments, i'm really thankful for all these awesome code optimizations, and i'm looking forward to try them and implement them if they work !!

but !!!!!

there are still tons of things to first or add first :rofl: 

just to add the boardsize to human readable function, i got to fix 3 different things (undefined error, then always reject, then reject message)

i was willing to go through this because there are direct benefits for the user :smile: 

but in this case we would want to rework all the code at once in a anti repetition code, so i'd rather consider this when we do it for all parts of the code that have same issue (spring cleaning :laughing: !!), and also when i have more knowledge to understand whats going on and fix it :100:

Dorus commented on game.js:
```diff
@@ -105,14 +105,28 @@ class Game {
+                && !clock.pause.pause_control["vacation-" + clock.black_player_id] && !clock.pause.pause_control["vacation-" + clock.white_player_id]) {
+                if (config.DEBUG) this.log("Pausing not allowed for unranked games. Resuming game.");
+                this.resumeGame();
+            }
```
Why do we have 3 if statements that all do `this.resumeGame();`? This could be one if statement.

wonderingabout replied:
because it depends if we want to resume for ranked games only or unranked only, my code is incomplete here indeed (need to add `&& !config.nopauseranked && !config.nopauseunranked`

Dorus replied:
You can still do it in one condition

wonderingabout replied:
well yes, but if we start optimizing code on this, then we need to do it for all similar ranked/unranked conditions (except UHMAEATs because they are different)

same comment for the nice functions you showed
they are nice indeed, but would need to do the same for all cases

wonderingabout replied:
my code is stupid copy pasting, but at least it works easily lol

your code proposals are awesome really, but they need time to do properly and to test
so my priority is adding new features and fixing bugs

enhancing already working things can be done later :+1:

Dorus replied:
Not all of them are this simple because most of the time the log messages are different. This one is especially easy, that's why i suggested to fix this one right away.

wonderingabout replied:
but then we enter the temptation circle lol 

and also, there are still many other things to fix, so as easy as it may be, i have to prioritirize bugs and fixes first :+1: 

so unlike you i consider this an enhancement, not a fix :laughing:

wonderingabout replied:
oh no actually here my code was correct 

- `if (config.nopause && !config.nopauseranked && !config.nopauseunranked `
- ` config.nopauseranked && this.state.ranked`
- ` config.nopauseunranked && (this.state.ranked == false)`

so now we pause either : 
- all games if --nopause (and not ranked/unranked of the same family, else only ranked and unranked of same family are taken into account, so maybe no pause only in ranked games, but allowed pause in unranked for example)
- only ranked games if --nopauseranked
- only unranked games if --nopauseunranked

nothing needs a fix here, only enhance existing code effiency :+1:

wonderingabout commented:
great !!! 

![nicedorus1](https://user-images.githubusercontent.com/38690718/52774282-24cc0700-303d-11e9-8c5a-10dd03d81c61.png)
![nicedorus2](https://user-images.githubusercontent.com/38690718/52774281-24cc0700-303d-11e9-897a-122bdb380437.png)

wonderingabout commented:
now will rename auto to automatic, for user easy read's sake : 

![autokomi](https://user-images.githubusercontent.com/38690718/52780789-564bcf00-304b-11e9-92cd-e18ef1baeb82.png)

wonderingabout commented:
is now !!

![isnow1](https://user-images.githubusercontent.com/38690718/52782257-e5a6b180-304e-11e9-8aac-4001f6df3f11.png)

wonderingabout commented:
@Dorus @roy7 

i made my first array today !!! (so i just realized "array" is the title of the chapter in javascript lesson where i stopped :rofl: )

managed to make this array by remembering what @Dorus told me last time

and also @Dorus i'm lazy to do separate warning messages, did this one though :1st_place_medal:

wonderingabout commented:
the warning message finally works now (the array was tricky because undefined for `--komiranked auto` and `--komiunranked auto` if these are not used

wonderingabout commented:
fixed and enhanced UHMAEAT to version 2.1 !!!

big credit goes to gedit "find and replace" feature for helping me do this so easily and reliably !!!
(it can change a block of many lines at the same time in all different parts of the file !!! 

changes : 

- fixed `undefined` in conn_log
- fixed canadian period time conn_log (period time for 10 stones is x seconds for all the 10 seconds)
- enhanced canadian period time code to make it more portable (now only one new property as compared to other time settings, easier to export !!!)


tested to work 10/10 !!

[uhmaeat2-1works3](https://user-images.githubusercontent.com/38690718/52804615-f40fc000-3084-11e9-935f-97d85c5baf40.png)
![uhmaeat2-1works2](https://user-images.githubusercontent.com/38690718/52804616-f40fc000-3084-11e9-891c-1ee8206e62c5.png)
![uhmaeat2-1works](https://user-images.githubusercontent.com/38690718/52804617-f40fc000-3084-11e9-9d8d-bae0d5db8774.png)

now ready to merge again @Dorus @roy7 
this is my final commit on this PR

wonderingabout commented:
![morescreenshots1](https://user-images.githubusercontent.com/38690718/52805532-1276bb00-3087-11e9-82ff-498cb8911a39.png)

![morescreenshots2](https://user-images.githubusercontent.com/38690718/52805531-1276bb00-3087-11e9-9bd3-73e3a6e12a45.png)
